Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp

```diff
@@ -145,15 +146,149 @@ s32 NWC24MakeUserID(u64* nwc24_id, u32 hollywood_id, u16 id_ctr, HardwareModel h
+
+  const Common::HttpRequest::Response response = m_http.Get(url);
+
+  if (!response)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to request data at {}", url);
+    return NWC24::WC24_ERR_SERVER;
+  }
+
+  // Check if the filesize is smaller than the header size.
+  if (response->size() < sizeof(NWC24::WC24File))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "File at {} is too small to be a valid file.", url);
+    return NWC24::WC24_ERR_BROKEN;
+  }
+
+  // Now we read the file
+  NWC24::WC24File wc24File;
+  std::memcpy(&wc24File, response->data(), sizeof(NWC24::WC24File));
+
+  std::vector<u8> temp_buffer(response->begin() + 320, response->end());
+
+  if (m_dl_list.IsEncrypted(entry_index))
+  {
+    NWC24::WC24PubkMod pubkMod = m_dl_list.GetWC24PubkMod(entry_index);
+
+    file_data = std::vector<u8>(response->size() - 320);
+
+    m_ios.GetIOSC().CryptOFB(pubkMod.aes_key, wc24File.iv, temp_buffer.data(), temp_buffer.size(),
+                             file_data.data());
+  }
+  else
+  {
+    file_data = std::move(temp_buffer);
+  }
+
+  NWC24::ErrorCode reply = IOS::HLE::NWC24::OpenVFF(m_dl_list.GetVFFPath(entry_index), content_name,
+                                                    m_ios.GetFS(), file_data);
+
+  return reply;
+}
+
+IPCReply NetKDRequestDevice::HandleNWC24DownloadNowEx(const IOCtlRequest& request)
+{
+  const s32 flags = static_cast<s32>(Memory::Read_U32(request.buffer_in));
+  // Nintendo converts the entry ID between a u32 and u16
+  // several times, presumably for alignment purposes.
+  // We'll skip past buffer_in+4 and keep the entry index as a u16.
+  const u16 entry_index = Memory::Read_U16(request.buffer_in + 6);
+  const s32 subtask_bitmask = static_cast<s32>(Memory::Read_U32(request.buffer_in + 8));
+
+  INFO_LOG_FMT(IOS_WC24,
+               "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - flags: {}, index: {}, bitmask: {}",
+               flags, entry_index, subtask_bitmask);
+
+  if (entry_index >= NWC24::NWC24Dl::MAX_ENTRIES)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "NET_KD_REQ: Entry index out of range.");
+    WriteReturnValue(NWC24::WC24_ERR_BROKEN, request.buffer_out);
+    return IPCReply(NWC24::WC24_ERR_BROKEN);
+  }
+
+  if (!m_dl_list.DoesEntryExist(entry_index))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "NET_KD_REQ: Requested entry does not exist in download list!");
+    WriteReturnValue(NWC24::WC24_ERR_NOT_FOUND, request.buffer_out);
+    return IPCReply(NWC24::WC24_ERR_NOT_FOUND);
+  }
+
+  // While in theory reply will always get initialized by KDDownload, things happen.
+  // Returning NWC24::WC24_ERR_BROKEN or anything that isn't OK will prompt the channel to fix the
+  // entry's data.
+  NWC24::ErrorCode reply = NWC24::WC24_ERR_BROKEN;
+
+  // Determine if we have subtasks to handle
+  if ((flags << 0x1d) < 0)
+  {
+    for (u8 subtask_id = 0; subtask_id < 32; subtask_id++)
+    {
+      // Check if we are done
+      if (((subtask_bitmask >> subtask_id) << 0x1f) >= 0)
```

Comment:
This confused me for a minute until I realized what it did... it's probably better expressed as `if (subtask_bitmask & (1 << subtask_id))`.